### ISSUE TITLE
Strictly typed functional templates

### ISSUE THREAD

**Issue by PavelPZ** [angular/angular#7339]
It would be great to have another possibility of template representation besides string. Inspired by React JSX/TSX, if there exists ```AngularJSX``` custom JSX factory (which creates JavaScript object tree, consumed by Angular as an alternative to string template) we can use JavaScript x Typescript template (see below). Both for inline and HTML templates. In the second case there could be e.g. ```templateJsxUrl='./teplates/myTemplate.jsx'``` component attribute:

## Pure JavaScript example:
```javascript
var AppComponent = function () {
  this.templateJsx = React.createElement("div", null, 
    React.createElement("h3", null, title), 
    this.items.map(function (it) { return this.displayItem(it); }), 
    this.okButton(this.doClick)
  );
  this.items = [1, 2, 3];
}
AppComponent.prototype.okButton = function (doClick) {
  return [
    AngularJSX.createElement("hr", null), 
    AngularJSX.createElement("div", {onClick: doClick}, " OK")];
};
AppComponent.prototype.displayItem = function (it) {
     AngularJSX.createElement("span", null, it);
};
```

## Strictly typed Typescript example 
if Angular accepts AngularJSX-like template representation, the rest is almost completely covered by Typescript 1.8., with all IDE benefits as refactoring, intellisence, syntax highlighting, auto-indent etc.

Developer writes:
```javascript
@Component({selector: 'my-app'}
class AppComponent {
  templateJsx = <div>
    <h3>{ title } </h3>
    { items.map(it => displayItem(it)) }
    { okButton(doClick) }
  </div>;

  okButton(doClick: React.MouseEventHandler) {
    return [<hr/>, <div onClick = { doClick }> OK</div >]
  }
  displayItem(it: any) {
    <span>{ it }</span>
  }
  items = [1,2,3];
  doClick () { alert('onClick'); }
}
...
```
Typescript 1.8 transpiler generates:
```javascript
...
this.templateJsx = AngularJSX.createElement("div", null, 
  AngularJSX.createElement("h3", null, title, 
  this.items.map(function (it) { return this.displayItem(it); }), 
  this.okButton(this.doClick)
);
...
```

For external template:

```javascript
@Component({selector: 'my-app'}
class AppComponent {
  templateJsxUrl = './teplates/myTemplate.jsx'
...
```

```AngularJSX.createElement``` encodes template to JavaScript object tree consumed by Angular as an string alternative.

**Comment by martinmcwhorter** [comment]
TL;DR - In the future TypeScript will be more extensible and will be able to type check HTML templates. No need for JSX/TSX in Angular.

**Comment by PavelPZ** [comment]
OK, but as wrote @Malekpour in #5131:
> value type of template will remain string which needs to be parsed and analysed at run-time, but value of templateJsx will be ready to use object or function

In other words, templateJsx can help JavaScript developers too.

**Comment by robwormald** [comment]
This is a duplicate of #5131, and the point has already been made that while the community is more than welcome to build such a setup, it's not something the team is interested in supporting.

**Comment by PavelPZ** [comment]
Ok, sorry, I understand now.
